### PR TITLE
Implemented async TimeSeriesClient

### PR DESCRIPTION
**Pull request by shaunsales**
Hi,

We use Redis TimeSeries quite extensively and are in need of an async version of the TimeSeriesClient. I've implemented it as a matching partial class with "Async" amended to the method names. The existing tests also have a matching async implementation.

Due to the parallel nature of the async tests, I had to ensure the uniqueness of Redis keys and their async disposal. This was implemented in the `AbstractTimeSeriesTest.cs` class as a couple of helper methods and the implementation of the `IAsyncLifetime` xUnit interface.

There is still further work to be done, but I felt this was a reasonable first cut to share. Hope it is useful.

Best regards,
Shaun

### CODE REVIEW

**Comment by DvirDukhan** [comment]
@shaunsales Thanks for this PR
I will review it in the next following days.
If you don't mind adding examples to the examples project, I will appreciate it.
Dvir

**Comment by shaunsales** [comment]
Thank you. Sure, I’ll do that in the next day or so.

File: NRedisTimeSeries/NRedisTimeSeries.csproj [hunk 1]
```diff
@@ -14,17 +14,10 @@
...
   </ItemGroup>
 
-  <ItemGroup>
-    <Folder Include="DataTypes\" />
-    <Folder Include="Commands\" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Remove="Interfaces\IValue.cs" />
-    <Compile Remove="Interfaces\ILabel.cs" />
-  </ItemGroup>
```
Comment (DvirDukhan) [inline NRedisTimeSeries/NRedisTimeSeries.csproj:29]: Leftovers?

File: NRedisTimeSeries/TimeSeriesClient.Async.cs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        public static async Task<IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, TimeSeriesTuple value)>> TimeSeriesMGetAsync(this IDatabase db, IReadOnlyCollection<string> filter, bool? withLabels = null)
+        {
+            var args = new List<object>();
+            args.AddWithLabels(withLabels);
+            AddFilters(args, filter);
+            return ParseMGetesponse(await db.ExecuteAsync(TS.MGET, args));
+        }
+
+        /// <summary>
+        ///
```
Comment (DvirDukhan) [inline NRedisTimeSeries/TimeSeriesClient.Async.cs:199]: Please update this once I update the normal API comments

File: NRedisTimeSeries/TimeSeriesClient.Async.cs [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        {
+            var args = new List<object>()
+            { key, fromTimeStamp.Value, toTimeStamp.Value };
+            args.AddCount(count);
+            args.AddAggregation(aggregation, timeBucket);
+            return ParseTimeSeriesTupleArray(await db.ExecuteAsync(TS.RANGE, args));
+        }
+
+        /// <summary>
+        ///
```
Comment (DvirDukhan) [inline NRedisTimeSeries/TimeSeriesClient.Async.cs:219]: Please update this once I update the normal API comments

File: NRedisTimeSeries/TimeSeriesClient.Async.cs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        /// </summary>
+        /// <param name="db">StackExchange.Redis IDatabase instance</param>
+        /// <param name="fromTimeStamp"> Start timestamp for the range query. - can be used to express the minimum possible timestamp.</param>
+        /// <param name="toTimeStamp">End timestamp for range query, + can be used to express the maximum possible timestamp.</param>
+        /// <param name="filter">A sequence of filters</param>
+        /// <param name="count">Optional: Maximum number of returned results per time-series.</param>
+        /// <param name="aggregation">Optional: Aggregation type</param>
+        /// <param name="timeBucket">Optional: Time bucket for aggregation in milliseconds</param>
+        /// <param name="withLabels">Optional: Include in the reply the label-value pairs that represent metadata labels of the time-series</param>
+        /// <returns></returns>
```
Comment (DvirDukhan) [inline NRedisTimeSeries/TimeSeriesClient.Async.cs:229]: Please update this once I update the normal API comments

File: NRedisTimeSeries/TimeSeriesClient.Async.cs [hunk 5]
```diff
@@ -0,0 +1,270 @@
+﻿using System;
```
Comment (DvirDukhan) [inline NRedisTimeSeries/TimeSeriesClient.Async.cs:1]: Please rename the file to `TimeSeriesClientAsync`

File: NRedisTimeSeries/TimeSeriesClient.Async.cs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        /// <returns>If the operation executed successfully</returns>
+        public static async Task<bool> TimeSeriesCreateRuleAsync(this IDatabase db, string sourceKey, TimeSeriesRule rule)
+        {
+            var args = new List<object> { sourceKey };
+            args.AddRule(rule);
+            return ParseBoolean(await db.ExecuteAsync(TS.CREATERULE, args));
+        }
+
+        /// <summary>
+        ///
```
Comment (DvirDukhan) [inline NRedisTimeSeries/TimeSeriesClient.Async.cs:156]: Please update this once I update the normal API comments

File: NRedisTimeSeries.Test/NRedisTimeSeries.Test.csproj [hunk 7]
```diff
@@ -2,25 +2,21 @@
...
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\NRedisTimeSeries\NRedisTimeSeries.csproj" />
   </ItemGroup>
 
-  <ItemGroup>
-    <Folder Include="TestDataTypes\" />
-    <Folder Include="TestAPI\" />
-  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (shaunsales) [reply NRedisTimeSeries.Test/NRedisTimeSeries.Test.csproj:25]: It's not necessary to specifically include folders in the `.csproj` if they are not empty, they will be picked up automatically by the IDE.

File: NRedisTimeSeries.Test/TestAPI/AbstractTimeSeriesTest.cs [hunk 8]
```diff
@@ -1,13 +1,71 @@
...
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
 using Xunit;
 
 namespace NRedisTimeSeries.Test.TestAPI
 {
-    public abstract class AbstractTimeSeriesTest: IClassFixture<RedisFixture>
+    public abstract class AbstractTimeSeriesTest : IClassFixture<RedisFixture>, IAsyncLifetime
```
[the remark this answers is not in the document]
Comment (shaunsales) [reply NRedisTimeSeries.Test/TestAPI/AbstractTimeSeriesTest.cs:10]: Thank you :)

File: NRedisTimeSeries.Test/TestAPI/AbstractTimeSeriesTest.cs [hunk 9]
```diff
@@ -1,13 +1,71 @@
...
     {
         protected internal RedisFixture redisFixture;
 
         protected internal AbstractTimeSeriesTest(RedisFixture redisFixture) => this.redisFixture = redisFixture;
 
+        private List<string> dbKeys = new List<string>();
+
+        protected internal string CreateKey([CallerMemberName] string memberName = "") => CreateKeys(1, memberName)[0];
+
+        protected internal string[] CreateKeys(int count, [CallerMemberName] string memberName = "")
```
[the remark this answers is not in the document]
Comment (shaunsales) [reply NRedisTimeSeries.Test/TestAPI/AbstractTimeSeriesTest.cs:20]: Sure, I'll make all the requested naming updates and submit.

**Comment by shaunsales** [comment]
> Thanks for the PR.
> I really like the addition in the abstract test class and I will modify the synchronized API test accordingly.
> 
> 1. Please use the implicit casting for timestamps. I added a comment over each explicit usage I found.
> 2. I will update the comments for the regular API, so please fill the gaps in the missing API comments when you can

Regarding the `TimeStamp` class, I found some issues with that particular type that I would like to address in a seperate pull request. In the meantime I'll revert things back to implicit casts.

File: NRedisTimeSeries/NRedisTimeSeries.csproj (hunk 1 above)
Comment (shaunsales) [reply NRedisTimeSeries/NRedisTimeSeries.csproj:29]: Yes, not sure what they were originally used for but sometimes Visual Studio will leave these in the `.csproj` after you scaffold the folder layout and move things around. They didn't seem to be used so I removed them.

**Comment by DvirDukhan** [comment]
@shaunsales 
1. I updated the missing comments in the synchronized API
2. Giving a second review for the abstract test class, I have this question: Why the dispose method needs to be async? you can just implement `IDisposable`, no?

**Comment by shaunsales** [comment]
> 
> 
> @shaunsales
> 
>     1. I updated the missing comments in the synchronized API
> 
>     2. Giving a second review for the abstract test class, I have this question: Why the dispose method needs to be async? you can just implement `IDisposable`, no?

Ok great, I'll update the comments in the async client.

Regarding the abstract test class, I felt that using xUnit's built-in async lifetime functionality would allow better extension in future, its pretty commonly used in xUnit tests, and can support both sync and async calls in Initialize/DisposeAsync - so really just for flexibility.

**Comment by shaunsales** [comment]
I've added async examples that match all of the sync examples, and fixed a bunch of typos and other formatting issues.

**Comment by DvirDukhan** [comment]
@shaunsales 
Thanks again for this great PR, I will review the new modifications soon.
Once I will merge it I will release version 1.1 and I want to mention your contribution once I'll tweet about this release. Do you prefer that I will share a link to your GitHub account, Twitter account, or both? I will need your Twitter account name in order to tag you.
Thanks

**Comment by shaunsales** [comment]
> 
> 
> @shaunsales
> Thanks again for this great PR, I will review the new modifications soon.
> Once I will merge it I will release version 1.1 and I want to mention your contribution once I'll tweet about this release. Do you prefer that I will share a link to your GitHub account, Twitter account, or both? I will need your Twitter account name in order to tag you.
> Thanks

Sure, happy to help - we love Redis and rely heavily on the TimeSeries module. I've got a few more suggestions/updates on the way once I have some free time. 

My Twitter is @shaunrsales - share whatever links you prefer.

I wanted to discuss the TimeStamp type you're using in the library. What's the best channel to contact you on?

**Comment by DvirDukhan** [comment]
@shaunsales 
dvir@redislabs.com

If it is related to this project, I prefer to have an open discussion as a GitHub issue, so we can involve the time series module team (@ashtul, @danni-m)

File: NRedisTimeSeries.Example/MGetExampleAsync.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using NRedisTimeSeries.DataTypes;
+using StackExchange.Redis;
+
+namespace NRedisTimeSeries.Example
+{
+    /// <summary>
+    /// Examples for NRedisTimeSeries API for MGET queries.
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MGetExampleAsync.cs:10]:
```suggestion
    /// Examples for NRedisTimeSeries async API for MGET queries.
```

File: NRedisTimeSeries.Example/MGetExampleAsync.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    /// </summary>
+    internal class MGetExampleAsync
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MGetExampleAsync.cs:12]:
```suggestion
    internal class MGetAsyncExample
```

File: NRedisTimeSeries.Example/MGetExampleAsync.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries MGET command with a filter.
+        /// The NRedisTimeSeries SimpleMGetExample returns and IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, TimeSeriesTuple value)> collection.
+        /// </summary>
+        public static async Task SimpleMGetExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MGetExampleAsync.cs:18]:
```suggestion
        public static async Task SimpleMGetAsyncExample()
```

File: NRedisTimeSeries.Example/MGetExampleAsync.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+                TimeSeriesTuple value = result.value;
+            }
+            redis.Close();
+        }
+
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries MGET command with a filter and WITHLABELS flag.
+        /// The NRedisTimeSeries SimpleMGetExample returns and IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, TimeSeriesTuple value)> collection.
+        /// </summary>
+        public static async Task MGetWithLabelsExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MGetExampleAsync.cs:37]:
```suggestion
        public static async Task MGetAsyncWithLabelsExample()
```

File: NRedisTimeSeries.Example/MRangeExampleAsync.cs [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using NRedisTimeSeries.Commands;
+using NRedisTimeSeries.DataTypes;
+using StackExchange.Redis;
+
+namespace NRedisTimeSeries.Example
+{
+    /// <summary>
+    /// Examples for NRedisTimeSeries API for MRANGE queries.
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MRangeExampleAsync.cs:11]:
```suggestion
    /// Examples for NRedisTimeSeries async API for MRANGE queries.
```

File: NRedisTimeSeries.Example/MRangeExampleAsync.cs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    /// </summary>
+    internal class MRangeExampleAsync
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MRangeExampleAsync.cs:13]:
```suggestion
    internal class MRangeAsyncExample
```

File: NRedisTimeSeries.Example/MRangeExampleAsync.cs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis and a filter.
+        /// NRedisTimeSeris MRange is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesMRange command returns an IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, IReadOnlyList<TimeSeriesTuple> values)>collection.
+        /// </summary>
+        public static async Task BasicMRangeExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MRangeExampleAsync.cs:21]:
```suggestion
        public static async Task BasicMRangeAsyncExample()
```

File: NRedisTimeSeries.Example/MRangeExampleAsync.cs [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+            redis.Close();
+        }
+
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis, a filter and the COUNT parameter.
+        /// NRedisTimeSeris MRange is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesMRange command returns an IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, IReadOnlyList<TimeSeriesTuple> values)>collection.
+        /// </summary>
+        public static async Task CountMRangeExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MRangeExampleAsync.cs:42]:
```suggestion
        public static async Task CountMRangeAsyncExample()
```

File: NRedisTimeSeries.Example/MRangeExampleAsync.cs [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            redis.Close();
+        }
+
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis, a filter and MIN aggregation.
+        /// NRedisTimeSeris MRange is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesMRange command returns an IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, IReadOnlyList<TimeSeriesTuple> values)>collection.
+        /// </summary>
+        public static async Task MRangeAggregationExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MRangeExampleAsync.cs:63]:
```suggestion
        public static async Task MRangeAsyncAggregationExample()
```

File: NRedisTimeSeries.Example/MRangeExampleAsync.cs [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+            redis.Close();
+        }
+
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis, a filter and WITHLABELS flag.
+        /// NRedisTimeSeris MRange is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesMRange command returns an IReadOnlyList<(string key, IReadOnlyList<TimeSeriesLabel> labels, IReadOnlyList<TimeSeriesTuple> values)>collection.
+        /// </summary>
+        public static async Task MRangeWithLabelsExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/MRangeExampleAsync.cs:84]:
```suggestion
        public static async Task MRangeAsyncWithLabelsExample()
```

File: NRedisTimeSeries.Example/RulesExampleAsync.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Threading.Tasks;
+using NRedisTimeSeries.Commands;
+using NRedisTimeSeries.DataTypes;
+using StackExchange.Redis;
+
+namespace NRedisTimeSeries.Example
+{
+    /// <summary>
+    /// Example for create and delete RedisTimeSeries rules.
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RulesExampleAsync.cs:10]:
```suggestion
    /// Example for async create and delete RedisTimeSeries rules.
```

File: NRedisTimeSeries.Example/RulesExampleAsync.cs [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    /// </summary>
+    internal class RulesExampleAsync
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RulesExampleAsync.cs:12]:
```suggestion
    internal class RulesAsyncExample
```

File: NRedisTimeSeries.Example/RulesExampleAsync.cs [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        /// <summary>
+        /// Example for create and delete RedisTimeSeries rules.
+        /// </summary>
+        public static async Task RulesCreateDeleteExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RulesExampleAsync.cs:17]:
```suggestion
        public static async Task RulesCreateDeleteAsyncExample()
```

File: NRedisTimeSeries.Example/RangeExampleAsync.cs [hunk 23]
```diff
@@ -0,0 +1,56 @@
+﻿using NRedisTimeSeries.Commands;
+using NRedisTimeSeries.DataTypes;
+using StackExchange.Redis;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace NRedisTimeSeries.Example
+{
+    /// <summary>
+    /// Examples for NRedisTimeSeries API for RANGE queries.
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RangeExampleAsync.cs:10]:
```suggestion
    /// Examples for NRedisTimeSeries async API for RANGE queries.
```

File: NRedisTimeSeries.Example/RangeExampleAsync.cs [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    /// </summary>
+    public class RangeExampleAsync
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RangeExampleAsync.cs:12]:
```suggestion
    public class RangeAsyncExample
```

File: NRedisTimeSeries.Example/RangeExampleAsync.cs [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis.
+        /// NRedisTimeSeris Range is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesRange command returns an IReadOnlyList<TimeSeriesTuple> collection.
+        /// </summary>
+        public static async Task DefaultRangeExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RangeExampleAsync.cs:20]:
```suggestion
        public static async Task DefaultRangeAsyncExample()
```

File: NRedisTimeSeries.Example/RangeExampleAsync.cs [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+            redis.Close();
+        }
+
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis, and the COUNT parameter.
+        /// NRedisTimeSeris Range is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesRange command returns an IReadOnlyList<TimeSeriesTuple> collection.
+        /// </summary>
+        public static async Task CountRangeExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RangeExampleAsync.cs:34]:
```suggestion
        public static async Task CountRangeAsyncExample()
```

File: NRedisTimeSeries.Example/RangeExampleAsync.cs [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+            redis.Close();
+        }
+
+        /// <summary>
+        /// Example for basic usage of RedisTimeSeries RANGE command with "-" and "+" as range boundreis, and MIN aggregation.
+        /// NRedisTimeSeris Range is expecting two TimeStamps objects as the range boundries.
+        /// In this case, the strings are implicitly casted into TimeStamp objects.
+        /// The TimeSeriesRange command returns an IReadOnlyList<TimeSeriesTuple> collection.
+        /// </summary>
+        public static async Task RangeAggregationExampleAsync()
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Example/RangeExampleAsync.cs:48]:
```suggestion
        public static async Task RangeAsyncAggregationExample()
```

**Comment by shaunsales** [comment]
All requested updates have been committed. I also updated the examples to promote the use of DateTime.UtcNow as we should be encouraging .NET developers away from using DateTime.Now for timestamp storage.

**Comment by DvirDukhan** [comment]
@shaunsales 
merged
I will release 1.1 soon, probably early next week. I have to close 2 issues first
Thanks a lot for this effort

**Pull request merged**